src/devices/lacrosse_ws7000.c

```diff
@@ -0,0 +1,282 @@
+LaCrosse WS7000/WS2500 weather sensors.
+
+- WS2500-19 brightness sensor
+- WS7000-20 meteo sensor (temperature/humidity/pressure)
+- WS7000-16 Rain Sensor
+- WS7000-15 wind sensor
+
+PWM 400 us / 800 us with fixed bit width of 1200 us.
+Messages are sent as nibbles (4 bits) with LSB sent first.
+A frame is composed of a preamble followed by nibbles (4 bits) separated by a 1-bit.
+
+Message Layout:
+
+    P P S A D..D X C
+
+- Preamble: 10x bit "0", bit "1"
+- Sensor Type:  Value 0..9 determing the sensor type
+  - 0 = WS7000-27/28 Thermo sensor ( interval 177s - Adr * 0.5s )
+  - 1 = WS7000-22/25 Thermo/Humidity sensor ( interval 177s - Adr * 0.5s )
+  - 2 = WS7000-16 Rain sensor ( interval 173s - Adr * 0.5s )
+  - 3 = WS7000-15 Wind sensor ( interval 169s - Adr * 0.5s )
+  - 4 = WS7000-20 Thermo/Humidity/Barometer sensor ( interval 165s - Adr * 0.5s )
+  - 5 = WS2500-19 Brightness sensor ( interval 161s - Adr * 0.5s )
+- Address:  Value 0..7 for the sensor address
+    In case of a negative temperature the MSB of the Address becomes "1"
+- Data:     Nibbles with BCD encoded sensor data values.
+- XOR:      Nibble holding XOR of the S ^ A ^ Data-nibbles
+- Checksum: Sum of all nibbles + 5 ( S + A + nibble(0) + .. + nibble(n) + XOR + 5 ) & 0xF
+
+*/
+
+#include "decoder.h"
+
+/// unstuff nibbles with 1-bit separator, returns number of sucessfully unstuffed nibbles.
+static unsigned unstuff_5to4(uint8_t *src, unsigned offset, unsigned len, uint8_t *dst)
+{
+    unsigned ret = 0;
+
+    while (len >= 5) {
+        uint16_t bits = (src[offset / 8] << 8) | src[(offset / 8) + 1];
+        bits >>= 11 - (offset % 8); // align 5 bits to LSB
+        if ((bits & 1) != 1)
+            break;
+        *dst++ = (bits >> 1) & 0xf;
+        ret += 1;
+        offset += 5;
+        len -= 5;
+    }
+
+    return ret;
+}
+
+static uint8_t reflect4(uint8_t x)
+{
+    x = (x & 0xCC) >> 2 | (x & 0x33) << 2;
+    x = (x & 0xAA) >> 1 | (x & 0x55) << 1;
+    return x;
+}
+
+static void reflect_nibbles(uint8_t message[], unsigned num_bytes)
+{
+    for (unsigned i = 0; i < num_bytes; ++i) {
+        message[i] = reflect4(message[i]);
+    }
+}
+
+static int lacrosse_ws7000_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    uint8_t const preamble_pattern[] = {0x01}; // 8 bits
+    uint8_t const data_size[] = {3, 6, 3, 6, 10, 7}; // data nibbles by sensor type
+
+    data_t *data;
+    uint8_t b[14] = {0}; // LaCrosse WS7000-20 meteo sensor: 14 nibbles
+
+    //bitbuffer_print(bitbuffer);
+
+    unsigned start_pos = bitbuffer_search(bitbuffer, 0, 0, preamble_pattern, 8) + 8;
+    if (start_pos >= bitbuffer->bits_per_row[0])
+        return 0;
+
+    unsigned max_bits = MIN(14 * 5, bitbuffer->bits_per_row[0] - start_pos);
+    unsigned len = unstuff_5to4(bitbuffer->bb[0], start_pos, max_bits, b);
+    if (len < 2) // at least type/addr nibbles
+        return 0;
+
+    //bitrow_printf(b, len * 8, "LaCrosse WS7000: ");
+
+    reflect_nibbles(b, len);
```

Comment:
Before the reflection we should reject based on bit length. So far we know  4 types of sensors and they all have different lengths. So it should be cheap to reject here.